Started to implement the energy-dependent 1D spectrum extraction

cosimoNigro opened the pull request:
**Description**

**What is the motivation for this pull request?**
This PR starts the implementation of the energy-dependent spectrum extraction.

**Briefly: what changes are done here?**
* For the moment I have modified the `SpectrumDatasetMaker` - actually the `MapDatasetMaker` from which it inherits - to use the information in `~gammapy.irf.RadMax2D` to create a different ON region for each energy bin and fill the array with the ON counts accordingly.

* The next step would be to modify the `ReflectedRegionsBackgroundMaker` to extract the OFF counts using the same strategy.

**If this is to address a Github issue, mention its number to create a link**
Related to Issue #3296.

**Status**
This is a draft, meant for discussion. No docs and no tests have been added yet.
But find here a snippet of how the energy-dependent spectrum extraction will look with these modifications (I stop at the `SpectrumDatasetMaker.run`).
```python
from astropy.coordinates import SkyCoord
from regions import PointSkyRegion
from gammapy.data import Observation
from gammapy.maps import MapAxis, RegionGeom
from gammapy.makers import SpectrumDatasetMaker
from gammapy.datasets import SpectrumDataset
import matplotlib.pyplot as plt

# load the observation
observation = Observation.read("$GAMMAPY_DATA/magic/rad_max/data/magic_dl3_run_05029747.fits")
# observation = Observation.read("$GAMMAPY_DATA/hess-dl3-dr1/data/hess_dl3_dr1_obs_id_020323.fits.gz")

# energy axes
energy_axis = MapAxis.from_energy_bounds(
    0.005, 50, nbin=28, per_decade=False, unit="TeV", name="energy"
)
energy_true_axis = MapAxis.from_energy_bounds(
    0.005, 50, nbin=20, per_decade=False, unit="TeV", name="energy_true"
)

# create a point-like geometry for the centre of the ON region
target_position = SkyCoord(ra=83.63, dec=22.01, unit="deg", frame="icrs")
on_center = PointSkyRegion(target_position)
geom = RegionGeom.create(region=on_center, axes=[energy_axis])

# spectrum dataset and its maker
dataset_empty = SpectrumDataset.create(
    geom=geom, energy_axis_true=energy_true_axis
)

dataset_maker = SpectrumDatasetMaker(
    containment_correction=False, selection=["counts", "exposure", "edisp", "rad_max"]
)

# try to run it and plot the counts
dataset = dataset_maker.run(
    dataset_empty.copy(name="trial"), observation
)
dataset.counts.plot(drawstyle="steps-mid", ls="-")
plt.grid(ls=":")
plt.show()
```

![Figure_1](https://user-images.githubusercontent.com/26409711/145576981-292a7ee3-63ff-42ad-be23-367bb6cf5475.png)

It looks in agreement with what I was obtaining with MARS and with my custom scripts
![run_5029747_counts_comparison](https://user-images.githubusercontent.com/26409711/145577401-e68a2b75-b557-4ec9-8876-78c902aeb013.png)

Thanks!

P.S.
It would help other testing if you merge https://github.com/gammapy/gammapy-data/pull/18.

maxnoe commented on gammapy/makers/map.py:
```diff
@@ -75,6 +78,35 @@ def make_counts(geom, observation):
+    def make_counts_rad_max(geom, observation):
+        """Extract the counts using for the ON region size the values in the `RAD_MAX_2D` table"""
+        if observation.rad_max is None:
+            raise TypeError("the IRF for this observation does not include a RAD_MAX_2D table")
```
I don't think this is a `TypeError`. `TypeError` is meant for cases like a function expecting an `int` but getting a `str`.

This is either a `ValueError` or maybe, also in regard to https://github.com/gammapy/gammapy/issues/3666, we should create a ` RequiredIRFMissing` or similar exception in gammapy.

adonath commented on gammapy/makers/map.py:
```diff
@@ -75,6 +78,35 @@ def make_counts(geom, observation):
         return counts
+    
+    @staticmethod
+    def make_counts_rad_max(geom, observation):
```
I agree the method make sense in terms of code organisation, but maybe not that much in terms of API? I think ideally the handling of energy dependent region sizes could be almost hidden completely to the user. E.g. like:

```python3
from gammapy.makers.utils import make_counts_rad_max

def make_counts(geom, observation):
      
      # TODO: Maybe add RegionGeom.is_energy_dependent or similar....
      if geom.is_region and isinstance(geom.region, PointSkyRegion):
          counts = make_counts_rad_max(geom, rad_max, events)
      else:
           counts.fill_events()

```

adonath commented on gammapy/makers/map.py:
```diff
@@ -35,7 +38,7 @@ class MapDatasetMaker(Maker):
 
     tag = "MapDatasetMaker"
-    available_selection = ["counts", "exposure", "background", "psf", "edisp"]
+    available_selection = ["counts", "exposure", "background", "psf", "edisp", "rad_max"]
```
The `rad_max` option here is not really part of the data reduction as the other quantities. I think the `selection` rather refers to "which" data to reduce, while the `rad_max` rather refers to "how" it is reduced. In practice it is actually not optional at all, because once the IRFs are point like and the `rad_max` info is present one has to conduct the data reduction this way. So I would completely remove the option and not expose it to the user.

adonath commented on gammapy/makers/map.py:
```diff
@@ -75,6 +78,35 @@ def make_counts(geom, observation):
+            counts.fill_events(observation.events)
+            counts_list.append(counts.data[0])
+
+        counts = RegionNDMap.from_geom(geom)
```
You can directly pass the data here, like `RegionNDMap.from_geom(geom=geom, data=np.array(counts_list))`

adonath replied:
Creating from the point geometry means the information on which `rad_max` was actually used will be lost later. Of course it can be re-created from the IRFs later, but those might not be accessible anymore. Technically one could work with a `CompoundSkyRegion` here, just to preserve the information at all. But this requires careful documentation to explain how the regions map to the energy bins. Maybe it creates more confusion than it helps...

adonath commented on gammapy/makers/map.py:
```diff
@@ -75,6 +78,35 @@ def make_counts(geom, observation):
+            energy_range = geom.axes["energy"].slice(i)
+            on_region_geom = RegionGeom(on_region, axes=[energy_range])
+            counts = Map.from_geom(on_region_geom)
+            counts.fill_events(observation.events)
```
Because of the lazy loading the access to `observation.events` will actually re-load the event from disk. To prevent this just introduce a local variable at the top of the method:

```python3
@staticmethod
def make_counts(geom, observation):
    events = observation.events
    ...
```

adonath reviewed (changes requested):
Thanks @cosimoNigro, I have left a few more comments. I think most importantly we should not expose any option to the users to choose the `rad_max`, because it is basically mandatory once the IRFs are defined. To leave the option for a later extension / refactoring with energy dependent region sizes on the `RegionGeom`, I would propose to keep the API changes minimal and just introduce `gammapy.makers.utils.make_counts_rad_max`, which is then called in `MapDatasetMaker.make_counts()`, based on the unique identification of the rad max case.

cosimoNigro commented:
Hello @adonath, @registerrier can you take a look now?
I implemented the energy-dependent extraction of the OFF counts using the `ReflectedRegionsBackgroundMaker` and added a test in `test_reduce`, testing the new functionalities. Let me know if you want me to add more tests and where.

This is how the energy-dependent spectrum extraction would look like. Basically identical to the fixed-energy one but passing a `PointSkyRegion` as centre of the ON region. I'll add a notebook in a follow-up PR.
```python
import astropy.units as u
from astropy.coordinates import SkyCoord, Angle
from regions import PointSkyRegion
from gammapy.data import Observation
from gammapy.maps import MapAxis, RegionGeom
from gammapy.makers import SpectrumDatasetMaker, ReflectedRegionsBackgroundMaker
from gammapy.datasets import SpectrumDataset
import matplotlib.pyplot as plt

# load the observation
observation = Observation.read("$GAMMAPY_DATA/magic/rad_max/data/magic_dl3_run_05029747.fits")

# energy axes
energy_axis = MapAxis.from_energy_bounds(
    0.005, 50, nbin=28, per_decade=False, unit="TeV", name="energy"
)
energy_true_axis = MapAxis.from_energy_bounds(
    0.005, 50, nbin=20, per_decade=False, unit="TeV", name="energy_true"
)

# create a point-like geometry for the centre of the ON region
target_position = SkyCoord(ra=83.63, dec=22.01, unit="deg", frame="icrs")
on_center = PointSkyRegion(target_position)
geom = RegionGeom.create(region=on_center, axes=[energy_axis])

# spectrum dataset and its maker (fills the ON counts)
dataset_empty = SpectrumDataset.create(
    geom=geom, energy_axis_true=energy_true_axis
)
dataset_maker = SpectrumDatasetMaker(
    containment_correction=False, selection=["counts", "exposure", "edisp"]
)
dataset = dataset_maker.run(
    dataset_empty.copy(name="trial"), observation
)

# reflected regions maker (fills the OFF counts)
bkg_maker = ReflectedRegionsBackgroundMaker()
dataset_on_off = bkg_maker.run(dataset, observation)

# try to plot them
dataset_on_off.counts.plot(drawstyle="steps-mid", ls="-", label="ON")
dataset_on_off.counts_off.plot(drawstyle="steps-mid", ls="-", label="OFF")
dataset_on_off.acceptance_off.plot(drawstyle="steps-mid", ls="-", label="acceptance")
plt.legend()
plt.grid(ls=":")
plt.show()
```
![Figure_1](https://user-images.githubusercontent.com/26409711/146740655-e5f7dd2f-2333-4c11-8413-047ad2b340b6.png)

maxnoe commented:
@cosimoNigro these on/off counts look wrong too me, there seems to be an issue. In the central bins you have roughly 3 times more off than on counts? 

And why are all off bins below 30 GeV empty but not the on counts?

cosimoNigro commented:
Hello @maxnoe 

> In the central bins you have roughly 3 times more off than on counts?

But you have to scale those for the acceptance

> And why are all off bins below 30 GeV empty but not the on counts?

In the low energy bins it fails to find anyOFF regions. That was a mistake from my side.
I stored a too large theta2 cut in the low energy bins, namely 0.2, which translates to theta = 0.45, so larger than the offset.
The pointing position is within the ON region and no symmetric OFF could be found.
Will fix this with the next data.

maxnoe commented:
@cosimoNigro Wait, does this code use different off regions (not just different sizes of off regions) for each energy bin?

That would be a very surprising behavior to me.

maxnoe commented:
Ah, and failure to find any off regions should probably result in an exception

cosimoNigro commented:
> does this code use different off regions (not just different sizes of off regions) for each energy bin?

Thanks Max, I did not think about this. 
Yes, for each estimated energy bins it is simply re-running the `ReflectedRegionsBackgroundFinder`, so trying to fit as many OFF regions as possible.
How can we force it to use the same number of OFF regions? (e.g. 1 or 3).
Specifying e.g. [`max_region_number=1`](https://github.com/gammapy/gammapy/blob/master/gammapy/makers/background/reflected.py#L75) will not work, as it will not place it symmetrically w.r.t. the ON position.

maxnoe commented:
@cosimoNigro I think this goes back to one of the initial discussions we had of implementing a different version of the `ReflectedRegionsBackgroundFinder` that will use the "classical" approach of distributing n off regions along

using `offset_angles = (np.arange(n_off_positions) + 1) * 2 * np.pi / (n_off_positions + 1)` ?

adonath commented:
@cosimoNigro and @maxnoe I think this could just be an additional option to the existing `ReflectedRegionsBackgroundMaker`, implemented independently of this PR. How does one handle exclusion regions in this approach?

maxnoe commented:
You remove positions that intersect with an exclusion Region.

registerrier commented:
I suppose it is unclear whether the `ReflectedRegionsBackgroundMaker` should support the `radmax2d` at all. At least the fixed RADMAX case should be supported.
 
For clarity, it might be simpler to have another background `Maker` for the wobble/symetric region. Otherwise, the arguments might get pretty complicated.

adonath commented on gammapy/makers/utils.py:
```diff
@@ -422,3 +424,161 @@ def make_theta_squared_table(
+    return rad_max
+
+
+def make_counts_rad_max(geom, observation):
```
One last minor comment here: we typically kept the signature of the `make_something` utility methods to take the IRF classes and not the observation objects. The idea I just that this way they can used a bit easier independently. The API relying on the `Obseervation` object is still available on the `Maker` class, like `MapDatasetMaker.make_off_counts(geom, observation)`. So maybe changee the signature here to take the `geom`, `rad_max` and `events` independently?

cosimoNigro replied:
Hello @adonath, I changed the function's signature from `make_counts_rad_max(geom, observation)` to `make_counts_rad_max(geom, rad_max, events)`.

adonath commented on gammapy/makers/utils.py:
```diff
@@ -422,3 +424,161 @@ def make_theta_squared_table(
+    return counts
+
+
+def make_counts_off_rad_max(
```
Same here maybe remove the `observation` container from the signature and rely on the individual classes instead..

cosimoNigro replied:
Changed the function's signature from `make_counts_off_rad_max(geom, observation)` to `make_counts_off_rad_max(geom, rad_max, events)`

adonath reviewed (changes requested):
Thanks @cosimoNigro, I have left two final minor comments on the signature of the `maker_...` utility methods, otherwise looks good to me know. The proposed changes to the reflected regions background maker can be implemented in a different PR.

cosimoNigro commented:
Hi @adonath, implemented the change in the function arguments you requested. Sorry for the delay.
Cheers!

registerrier commented on gammapy/makers/utils.py:
```diff
@@ -422,3 +424,161 @@ def make_theta_squared_table(
+    counts_list = []
+
+    # create and fill a map per each energy bin, fetch the counts
+    for i, rad in enumerate(rad_max):
```
Maybe a comment here:
since RAD_MAX is determined independently of any projection, the region based selection is not a necessity. It might be more efficient here to perform the filtering on the `EventList` directly.

maxnoe replied:
Yes, just using `coordinate.separation(source_position) < rad_max` should be more efficient

cosimoNigro replied:
But the `RegionGeom` serves also the purpose of filtering by energy, if you look at it, it is defined by a single energy bin
If I filter your way then I have to filter also by energy, right?

registerrier commented on gammapy/makers/utils.py:
```diff
@@ -422,3 +424,161 @@ def make_theta_squared_table(
+    counts : `~gammapy.maps.RegionNDMap`
+        Counts vs estimated energy extracted from the ON region.
+    """
+    rad_max = get_rad_max_vs_energy(rad_max, events.pointing_radec, geom)
```
A question here. Is the rad_max determination usually done directly from the event energy or from the bin energy? i.e. is it expected that different energy binning yield different select event lists?

maxnoe replied:
Energy binning of what? I'd say you look up the cut value based on the event energy, like we do here: 

https://github.com/cta-observatory/pyirf/blob/3f9cfba672c8a25eb5094d25acf411c6b2104112/pyirf/cuts.py#L104

registerrier reviewed (approved):
Thanks @cosimoNigro ! This looks good.

I left two questions inline but they can be addressed later.

registerrier commented:
Thanks @cosimoNigro ! Let's merge now and rediscuss the other aspects in the follow-up PRs.
